### PR TITLE
[DNM] raftstore: divide localstate into raftstate and applystate

### PR DESCRIPTION
**Pull request by BusyJay**
Because we disable wal when write apply result, it may lose data when shutdown. It's safe when apply related meta and data are in one batch. So we need to divide localstate into raftstate and applystate.

@ngaut @siddontang @disksing @zhangjinpeng1987 PTAL

### CODE REVIEW

File: src/raftstore/store/peer_storage.rs [hunk 1]
```diff
@@ -691,13 +715,13 @@ mod test {
...
         store.rl().append(&mut ctx, &ents[1..]).expect("");
-        ctx.local_state.mut_truncated_state().set_index(ents[0].get_index());
-        ctx.local_state.mut_truncated_state().set_term(ents[0].get_term());
-        ctx.local_state.set_applied_index(ents.last().unwrap().get_index());
-        ctx.save(store.rl().get_region_id()).unwrap();
+        ctx.apply_state.mut_truncated_state().set_index(ents[0].get_index());
+        ctx.apply_state.mut_truncated_state().set_term(ents[0].get_term());
+        ctx.apply_state.set_applied_index(ents.last().unwrap().get_index());
+        ctx.save_apply(store.rl().get_region_id()).unwrap();
         store.rl().engine.write(ctx.wb).expect("");
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer_storage.rs]: should use wirte_without_wal

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -465,21 +484,26 @@ impl PeerStorage {
...
+        if ctx.raft_state != self.raft_state {
+            try!(ctx.save_raft(region_id));
+        }
+
+        if ctx.apply_state != self.apply_state {
+            try!(ctx.save_apply(region_id));
         }
 
         if !ctx.wb.is_empty() {
             try!(self.engine.write(ctx.wb));
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer_storage.rs]: should use write_without_wal

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -701,6 +701,12 @@ impl Peer {
 
         if data.is_empty() {
             // when a peer become leader, it will send an empty entry.
+            let wb = WriteBatch::new();
+            let mut state = self.storage.rl().apply_state.clone();
+            state.set_applied_index(index);
+            try!(wb.put_msg(&keys::apply_state_key(self.region_id), &state));
+            try!(self.engine.write(wb));
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer.rs]: should use write_without_wal

File: src/raftstore/store/peer_storage.rs [hunk 4]
```diff
@@ -465,21 +484,26 @@ impl PeerStorage {
...
 
-        if ctx.local_state != self.local_state {
-            try!(ctx.save(region_id));
+        if ctx.raft_state != self.raft_state {
+            try!(ctx.save_raft(region_id));
+        }
+
+        if ctx.apply_state != self.apply_state {
+            try!(ctx.save_apply(region_id));
         }
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer_storage.rs]: should not save apply state with wal

**Comment by ngaut** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**